language/stackless-bytecode/tree_heap/src/translator.rs

```diff
@@ -587,34 +637,82 @@ impl<'a> ModuleTranslator<'a> {
         }
         if inline {
             format!(
-                "procedure {{:inline 1}} {}_inline ({}) returns ({})",
+                "procedure {{:inline 1}} {} ({}) returns ({})",
                 fun_name, args, rets
             )
         } else {
-            format!("procedure {} ({}) returns ({})", fun_name, args, rets)
+            format!(
+                "procedure {}_verify ({}) returns ({})",
+                fun_name, args, rets
+            )
         }
     }
 
-    pub fn generate_function_body(
+    // return string for body of verify function, which is just a call to the
+    // inline version of the function.
+    pub fn generate_verify_function_body(
         &self,
         idx: usize,
-        inline: bool,
         arg_names: &Option<Vec<String>>,
     ) -> String {
+        let fun_name = self.function_name_from_definition_index(idx);
+        let function_def = &self.module.function_defs()[idx];
+        let function_handle = self.module.function_handle_at(function_def.function);
+        let function_signature = self.module.function_signature_at(function_handle.signature);
+        let mut args = String::new(); // vector of ", argname"
+        let mut rets = String::new(); // vector of ", argname"
+                                      // return values are: <mutable references>, <actual returns>
+        for (i, arg_type) in function_signature.arg_types.iter().enumerate() {
+            if i > 0 {
+                args.push_str(", ");
+            }
+            args.push_str(&self.get_arg_name(i, arg_names).to_string());
+            // collect mutable reference return values
+            if arg_type.is_mutable_reference() {
```

Comment:
Why are we returning mutable references here? We shouldn't need to do it in this model right? And the signature generated also doesn't return arguments that are mutable references.

Reply:
You may be right - I was a bit confused about this.  I suggest we go ahead and merge this version and then you can fix it up.